src/transformStateWithClones.js

```diff
@@ -8,6 +8,43 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  let newState = {
+    ...state,
+  };
+  const resArray = [];
+
+  for (const action of actions) {
+    newState = {
+      ...newState,
+    };
+
+    switch (action['type']) {
+      case 'addProperties':
+        Object.assign(newState, action['extraData']);
+        resArray.push(newState);
+        break;
+
+      case 'removeProperties':
+        for (const key of action['keysToRemove']) {
+          delete newState[key];
+        }
+        resArray.push(newState);
+
+        break;
+
+      case 'clear':
+        for (const key in newState) {
+          delete newState[key];
+        }
+        resArray.push(newState);
+        break;
+
+      default:
+        throw new Error('action is not supported');
+    }
+  }
```

Comment:
you could make it shorter
```suggestion
  for (const action of actions) {
    switch (action['type']) {
      case 'addProperties':
        Object.assign(newState, action['extraData']);
        break;

      case 'removeProperties':
        for (const key of action['keysToRemove']) {
          delete newState[key];
        }
        break;

      case 'clear':
        for (const key in newState) {
          delete newState[key];
        }
        break;

      default:
        throw new Error('action is not supported');
    }

    resArray.push({ ...newState }); // pushing a copy of a current state after mutations above
  }
```